File: api/src/models/User.model.ts
import mongoose, { Document, Schema, Types } from 'mongoose';

export interface DailyActivity {
  date: Date;
  steps: number;
  distance: number;
}

export interface UserDoc extends Document {
  name: string;
  email: string;
  _id: string;
  walletAddress: string | null;
  activityData: DailyActivity[];
  createdAt: Date;
  updatedAt: Date;
}

// Create the User schema
const UserSchema: Schema = new Schema(
  {
    _id: { type: String, required: true, unique: true },
    name: { type: String, required: true },
    email: { type: String, required: true, unique: true },
    walletAddress: { type: String, default: null },
    activityData: [
      {
        date: { type: Date, required: true },
        steps: { type: Number, default: 0 },
        distance: { type: Number, default: 0 },
      }
    ]
  },
  { timestamps: true }
);

UserSchema.index({ email: 1 });
UserSchema.index({ walletAddress: 1 });
UserSchema.index({ 'activityData.date': 1 });

const User = mongoose.model<UserDoc>('User', UserSchema);

export default User;

File: app/lib/models/UserModel.dart
// ignore_for_file: file_names

class DailyActivity {
  final DateTime date;
  final int steps;
  final double distance;

  DailyActivity({
    required this.date,
    this.steps = 0,
    this.distance = 0.0,
  });

  factory DailyActivity.fromJson(Map<String, dynamic> json) {
    return DailyActivity(
      date: DateTime.parse(json['date']),
      steps: json['steps'] ?? 0,
      distance: (json['distance'] ?? 0.0).toDouble(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'date': date.toIso8601String(),
      'steps': steps,
      'distance': distance,
    };
  }
}

class UserModel {
  final String id;
  final String name;
  final String email;
  final String? walletAddress;
  final List<DailyActivity> activityData;
  final DateTime createdAt;
  final DateTime updatedAt;

  UserModel({
    required this.id,
    required this.name,
    required this.email,
    this.walletAddress,
    this.activityData = const [],
    required this.createdAt,
    required this.updatedAt,
  });

  factory UserModel.fromJson(Map<String, dynamic> json) {
    return UserModel(
      id: json['_id'],
      name: json['name'],
      email: json['email'],
      walletAddress: json['walletAddress'],
      activityData: (json['activityData'] as List)
          .map((item) => DailyActivity.fromJson(item))
          .toList(),
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: DateTime.parse(json['updatedAt']),
    );
  }
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'email': email,
      'walletAddress': walletAddress,
      'activityData': activityData.map((item) => item.toJson()).toList(),
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
  }

  @override
  String toString() {
    return 'UserModel{id: $id, name: $name, email: $email, walletAddress: $walletAddress, activityData: $activityData, createdAt: $createdAt, updatedAt: $updatedAt}';
  }
}
